internal/utils/azure/azure.go

```diff
@@ -53,94 +75,119 @@ func NewAzureAPI() (*AzureAPI, error) {
 	pager := c.client.NewListForScopePager(scope, &armauthorization.DenyAssignmentsClientListForScopeOptions{
 		Filter: filter,
 	})
 
-	var denyAssignments []*armauthorization.DenyAssignment
-	for pager.More() {
-		nextResult, err := pager.NextPage(context.TODO())
-		if err != nil {
-			return nil, fmt.Errorf("failed to get next page of results: %w", err)
-		}
-		if nextResult.Value != nil {
-			denyAssignments = append(denyAssignments, nextResult.Value...)
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+		for pager.More() {
+			nextResult, err := pager.NextPage(c.ctx)
+			if err != nil {
+				ch <- fmt.Errorf("failed to get next page of results: %w", err)
+			}
+			if nextResult.Value != nil {
+				denyAssignments = append(denyAssignments, nextResult.Value...)
+			}
 		}
+		ch <- nil
+	}()
+
+	select {
+	case err := <-ch:
+		return denyAssignments, err
+	case <-c.ctx.Done():
+		return denyAssignments, fmt.Errorf("context cancelled")
 	}
-
-	return denyAssignments, nil
 }
 
 // AzureRoleAssignmentsClient is a facade over the Azure role assignments client. Exists to make our
 // code easier to test (it handles paging).
 type AzureRoleAssignmentsClient struct {
+	ctx    context.Context
 	client *armauthorization.RoleAssignmentsClient
 }
 
 // NewAzureRoleAssignmentsClient creates a new AzureRoleAssignmentsClient (our facade client) from a
 // client from the Azure SDK.
-func NewAzureRoleAssignmentsClient(azClient *armauthorization.RoleAssignmentsClient) *AzureRoleAssignmentsClient {
+func NewAzureRoleAssignmentsClient(ctx context.Context, azClient *armauthorization.RoleAssignmentsClient) *AzureRoleAssignmentsClient {
 	return &AzureRoleAssignmentsClient{
+		ctx:    ctx,
 		client: azClient,
 	}
 }
 
 // GetRoleAssignmentsForScope gets all the role assignments matching a scope and an optional filter.
 func (c *AzureRoleAssignmentsClient) GetRoleAssignmentsForScope(scope string, filter *string) ([]*armauthorization.RoleAssignment, error) {
+	var roleAssignments []*armauthorization.RoleAssignment
 	pager := c.client.NewListForScopePager(scope, &armauthorization.RoleAssignmentsClientListForScopeOptions{
 		Filter: filter,
 	})
 
-	var roleAssignments []*armauthorization.RoleAssignment
-	for pager.More() {
-		nextResult, err := pager.NextPage(context.TODO())
-		if err != nil {
-			return nil, fmt.Errorf("failed to get next page of results: %w", err)
-		}
-		if nextResult.Value != nil {
-			roleAssignments = append(roleAssignments, nextResult.Value...)
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+		for pager.More() {
+			nextResult, err := pager.NextPage(c.ctx)
+			if err != nil {
+				ch <- fmt.Errorf("failed to get next page of results: %w", err)
+			}
+			if nextResult.Value != nil {
+				roleAssignments = append(roleAssignments, nextResult.Value...)
+			}
 		}
+		ch <- nil
+	}()
+
+	select {
+	case err := <-ch:
+		return roleAssignments, err
+	case <-c.ctx.Done():
+		return roleAssignments, fmt.Errorf("context cancelled")
```

Comment:
same as the comment above, wouldnt this end up returning the context cancelled error along with all the role assignments in the happy case? imo it might be better to add the case where we pull the nil from the channel and return `roleAssignments, nil`